clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/JsonTemplatesProducer.kt

```diff
@@ -0,0 +1,252 @@
+                builder.name(templateName)
+                builder.presenter(templatesPresenter)
+                addJsonArgumentsToBuilder(builder, json.getJSONObject("arguments"))
+                builder.build()
+            }
+
+            CustomTemplateType.FUNCTION -> {
+                if (functionsPresenter == null) {
+                    throw CustomTemplateException("JSON definition contains a function definition and a function presenter is required")
+                }
+
+                val isVisual = json.getBoolean("isVisual")
+
+                val builder = CustomTemplate.FunctionBuilder(isVisual)
+                builder.name(templateName)
+                builder.presenter(functionsPresenter)
+                addJsonArgumentsToBuilder(builder, json.getJSONObject("arguments"))
+                builder.build()
+            }
+        }
+    }
+
+    private fun addJsonArgumentsToBuilder(
+        builder: CustomTemplate.TemplateBuilder,
+        json: JSONObject
+    ) {
+        for (key in json.keys()) {
+            val argumentJson = json.getJSONObject(key)
+            val typeString = argumentJson.getString("type")
+            if (typeString == "object") {
+                builder.mapArgument(key, jsonArgToMap(argumentJson.getJSONObject("value")))
+                continue
+            }
+
+            val argumentType = TemplateArgumentType.fromString(typeString)
+                ?: throw CustomTemplateException("Unsupported argument type: \"$typeString\"")
+
+            when (argumentType) {
+                BOOLEAN -> {
+                    val value = argumentJson.getBoolean("value")
+                    builder.booleanArgument(key, value)
+                }
+
+                NUMBER -> {
+                    val value = argumentJson.getDouble("value")
+                    builder.doubleArgument(key, value)
+                }
+
+                STRING -> {
+                    val value = argumentJson.getString("value")
+                    builder.stringArgument(key, value)
+                }
+
+                FILE -> {
+                    if (argumentJson.has("value")) {
+                        throw CustomTemplateException("File arguments should not specify a value. Remove value from argument: \"$key\"")
+                    }
+
+                    builder.fileArgument(key)
+                }
+
+                ACTION -> {
+                    if (argumentJson.has("value")) {
+                        throw CustomTemplateException("Action arguments should not specify a value. Remove value from argument: \"$key\"")
+                    }
+
+                    builder.actionArgument(key)
+                }
+            }
+        }
+    }
+
+    private fun addJsonArgumentsToBuilder(
+        builder: CustomTemplate.FunctionBuilder,
+        json: JSONObject
+    ) {
+        for (key in json.keys()) {
+            val argumentJson = json.getJSONObject(key)
```

Comment:
might throw exception, make sure it is handled (wont throw ever/catch it)

Reply:
It is caught in the try-catch block of `defineTemplates` and re-thrown with a `CustomTemplateException`. I've added comments to all methods that throw to make that more clear.